presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,947 @@
+                newTableName = newTableName.toUpperCase(ENGLISH);
+            }
+            execute(connection, sql);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+    }
+
+    private String getColumnDefinitionSql(ColumnMetadata column, String columnName)
+    {
+        StringBuilder sb = new StringBuilder()
+                .append(quoted(columnName))
+                .append(" ");
+        if (column.isNullable()) {
+            sb.append("Nullable(").append(toWriteMapping(column.getType())).append(")");
+        }
+        else {
+            sb.append(toWriteMapping(column.getType()));
+        }
+        return sb.toString();
+    }
+
+    protected String createTableSql(RemoteTableName remoteTableName, List<String> columns, ConnectorTableMetadata tableMetadata)
+    {
+        ImmutableList.Builder<String> tableOptions = ImmutableList.builder();
+        Map<String, Object> tableProperties = tableMetadata.getProperties();
+        ClickHouseEngineType engine = ClickHouseTableProperties.getEngine(tableProperties);
+        tableOptions.add("ENGINE = " + engine.getEngineType());
+        if (engine == ClickHouseEngineType.MERGETREE && formatProperty(ClickHouseTableProperties.getOrderBy(tableProperties)).equals(Optional.empty())) {
+            // order_by property is required
+            throw new PrestoException(INVALID_TABLE_PROPERTY,
+                    format("The property of %s is required for table engine %s", ClickHouseTableProperties.ORDER_BY_PROPERTY, engine.getEngineType()));
+        }
+        formatProperty(ClickHouseTableProperties.getOrderBy(tableProperties)).ifPresent(value -> tableOptions.add("ORDER BY " + value));
+        formatProperty(ClickHouseTableProperties.getPrimaryKey(tableProperties)).ifPresent(value -> tableOptions.add("PRIMARY KEY " + value));
+        formatProperty(ClickHouseTableProperties.getPartitionBy(tableProperties)).ifPresent(value -> tableOptions.add("PARTITION BY " + value));
+        ClickHouseTableProperties.getSampleBy(tableProperties).ifPresent(value -> tableOptions.add("SAMPLE BY " + value));
+
+        return format("CREATE TABLE %s (%s) %s", quoted(remoteTableName), join(", ", columns), join(" ", tableOptions.build()));
+    }
+
+    /**
+     * format property to match ClickHouse create table statement
+     *
+     * @param prop property will be formatted
+     * @return formatted property
+     */
+    private Optional<String> formatProperty(List<String> prop)
+    {
+        if (prop == null || prop.isEmpty()) {
+            return Optional.empty();
+        }
+        else if (prop.size() == 1) {
+            // only one column
+            return Optional.of(prop.get(0));
+        }
+        else {
+            // include more than one columns
+            return Optional.of("(" + String.join(",", prop) + ")");
+        }
+    }
+
+    private String toWriteMapping(Type type)
```

Comment:
`toWriteMapping` is used twice in the same function. Would be nice to inline it. We could consider keeping a map, and just make a call?